File: backend/modules/playback.py
import asyncio
import spotipy

from database.db import query
from modules.room import get_room_by_code


def _get_spotify_client(access_token: str):
    return spotipy.Spotify(auth=access_token, requests_timeout=20)


async def add_to_queue(room_code: str, track: dict, added_by: str):
    try:
        room = await get_room_by_code(room_code)

        if not room:
            raise Exception('Room not found')

        pos_result = await query(
            """
            SELECT COALESCE(MAX(position), -1) as max_pos
            FROM queue_items
            WHERE room_id = $1
            """,
            [room['id']],
        )

        next_position = (pos_result.rows[0]['max_pos'] or -1) + 1

        await query(
            """
            INSERT INTO queue_items
            (room_id, track_uri, track_name, artist_name, album_name, duration_ms, added_by, position)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
            """,
            [
                room['id'],
                track.get('uri'),
                track.get('name'),
                track.get('artists'),
                track.get('album'),
                track.get('durationMs'),
                added_by,
                next_position,
            ],
        )

        return await get_queue(room_code)
    except Exception as exc:
        print('Error adding to queue:', exc)
        raise


async def get_queue(room_code: str):
    try:
        room = await get_room_by_code(room_code)

        if not room:
            return []

        result = await query(
            """
            SELECT id, track_uri, track_name, artist_name, album_name,
                   duration_ms, added_by, position, added_at
            FROM queue_items
            WHERE room_id = $1
            ORDER BY position
            """,
            [room['id']],
        )

        return result.rows
    except Exception as exc:
        print('Error getting queue:', exc)
        return []


async def remove_from_queue(room_code: str, queue_item_id: int):
    try:
        room = await get_room_by_code(room_code)

        if not room:
            raise Exception('Room not found')

        await query(
            """
            DELETE FROM queue_items
            WHERE room_id = $1 AND id = $2
            """,
            [room['id'], queue_item_id],
        )

        await query(
            """
            UPDATE queue_items
            SET position = subquery.new_position
            FROM (
              SELECT id, ROW_NUMBER() OVER (ORDER BY position) - 1 as new_position
              FROM queue_items
              WHERE room_id = $1
            ) as subquery
            WHERE queue_items.id = subquery.id
            """,
            [room['id']],
        )

        return await get_queue(room_code)
    except Exception as exc:
        print('Error removing from queue:', exc)
        raise


async def clear_queue(room_code: str):
    try:
        room = await get_room_by_code(room_code)

        if not room:
            raise Exception('Room not found')

        await query(
            """
            DELETE FROM queue_items WHERE room_id = $1
            """,
            [room['id']],
        )
    except Exception as exc:
        print('Error clearing queue:', exc)


async def transfer_playback(access_token: str, device_id: str):
    try:
        spotify = _get_spotify_client(access_token)
        await asyncio.to_thread(spotify.transfer_playback, device_id=device_id, force_play=False)

        print(f"Transferred playback to device {device_id}")
        return True
    except Exception as exc:
        print('Error transferring playback:', exc)
        raise


async def play(access_token: str, device_id: str, track_uri: str = None, position_ms: int = 0):
    try:
        spotify = _get_spotify_client(access_token)
        if track_uri:
            await asyncio.to_thread(
                spotify.start_playback,
                device_id=device_id,
                uris=[track_uri],
                position_ms=position_ms,
            )
        else:
            await asyncio.to_thread(spotify.start_playback, device_id=device_id)

        return True
    except Exception as exc:
        print('Error starting playback:', exc)
        raise


async def pause(access_token: str, device_id: str = None):
    try:
        spotify = _get_spotify_client(access_token)
        await asyncio.to_thread(spotify.pause_playback, device_id=device_id)

        return True
    except Exception as exc:
        print('Error pausing playback:', exc)
        raise


async def skip_to_next(access_token: str, device_id: str = None):
    try:
        spotify = _get_spotify_client(access_token)
        await asyncio.to_thread(spotify.next_track, device_id=device_id)

        return True
    except Exception as exc:
        print('Error skipping to next:', exc)
        raise


async def skip_to_previous(access_token: str, device_id: str = None):
    try:
        spotify = _get_spotify_client(access_token)
        await asyncio.to_thread(spotify.previous_track, device_id=device_id)

        return True
    except Exception as exc:
        print('Error skipping to previous:', exc)
        raise


async def seek(access_token: str, position_ms: int, device_id: str = None):
    try:
        spotify = _get_spotify_client(access_token)
        await asyncio.to_thread(spotify.seek_track, position_ms, device_id=device_id)

        return True
    except Exception as exc:
        print('Error seeking:', exc)
        raise


async def get_current_playback(access_token: str):
    try:
        spotify = _get_spotify_client(access_token)
        return await asyncio.to_thread(spotify.current_playback)
    except Exception as exc:
        print('Error getting current playback:', exc)
        raise


async def search_tracks(access_token: str, query: str, limit: int = 20):
    try:
        spotify = _get_spotify_client(access_token)
        data = await asyncio.to_thread(spotify.search, q=query, type='track', limit=limit)

        return [
            {
                'uri': track['uri'],
                'id': track['id'],
                'name': track['name'],
                'artists': ', '.join(artist['name'] for artist in track['artists']),
                'album': track['album']['name'],
                'albumArt': (track['album']['images'][0]['url'] if track['album']['images'] else None),
                'durationMs': track['duration_ms'],
                'previewUrl': track['preview_url'],
            }
            for track in data['tracks']['items']
        ]
    except Exception as exc:
        print('Error searching tracks:', exc)
        raise

File: backend/src/modules/playback.js
import axios from 'axios';
import { query } from '../database/db.js';
import { getRoomByCode } from './room.js';

export async function addToQueue(roomCode, track, addedBy) {
  try {
    const room = await getRoomByCode(roomCode);
    
    if (!room) {
      throw new Error('Room not found');
    }

    // Get current max position
    const posResult = await query(
      `SELECT COALESCE(MAX(position), -1) as max_pos
       FROM queue_items
       WHERE room_id = $1`,
      [room.id]
    );

    const nextPosition = posResult.rows[0].max_pos + 1;

    await query(
      `INSERT INTO queue_items
       (room_id, track_uri, track_name, artist_name, album_name, duration_ms, added_by, position)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
      [
        room.id,
        track.uri,
        track.name,
        track.artists,
        track.album,
        track.durationMs,
        addedBy,
        nextPosition
      ]
    );

    return await getQueue(roomCode);
  } catch (error) {
    console.error('Error adding to queue:', error);
    throw error;
  }
}

export async function getQueue(roomCode) {
  try {
    const room = await getRoomByCode(roomCode);
    
    if (!room) {
      return [];
    }

    const result = await query(
      `SELECT id, track_uri, track_name, artist_name, album_name,
              duration_ms, added_by, position, added_at
       FROM queue_items
       WHERE room_id = $1
       ORDER BY position`,
      [room.id]
    );

    return result.rows;
  } catch (error) {
    console.error('Error getting queue:', error);
    return [];
  }
}

export async function removeFromQueue(roomCode, queueItemId) {
  try {
    const room = await getRoomByCode(roomCode);
    
    if (!room) {
      throw new Error('Room not found');
    }

    await query(
      `DELETE FROM queue_items
       WHERE room_id = $1 AND id = $2`,
      [room.id, queueItemId]
    );

    // Reorder remaining items
    await query(
      `UPDATE queue_items
       SET position = subquery.new_position
       FROM (
         SELECT id, ROW_NUMBER() OVER (ORDER BY position) - 1 as new_position
         FROM queue_items
         WHERE room_id = $1
       ) as subquery
       WHERE queue_items.id = subquery.id`,
      [room.id]
    );

    return await getQueue(roomCode);
  } catch (error) {
    console.error('Error removing from queue:', error);
    throw error;
  }
}

export async function clearQueue(roomCode) {
  try {
    const room = await getRoomByCode(roomCode);
    
    if (!room) {
      throw new Error('Room not found');
    }

    await query(
      `DELETE FROM queue_items WHERE room_id = $1`,
      [room.id]
    );
  } catch (error) {
    console.error('Error clearing queue:', error);
  }
}

export async function transferPlayback(accessToken, deviceId) {
  try {
    await axios.put(
      'https://api.spotify.com/v1/me/player',
      {
        device_ids: [deviceId],
        play: false
      },
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log(`Transferred playback to device ${deviceId}`);
    return true;
  } catch (error) {
    console.error('Error transferring playback:', error.response?.data || error.message);
    throw error;
  }
}

export async function play(accessToken, deviceId, trackUri = null, positionMs = 0) {
  try {
    const body = trackUri ? { uris: [trackUri], position_ms: positionMs } : {};
    
    await axios.put(
      `https://api.spotify.com/v1/me/player/play${deviceId ? `?device_id=${deviceId}` : ''}`,
      body,
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return true;
  } catch (error) {
    console.error('Error starting playback:', error.response?.data || error.message);
    throw error;
  }
}

export async function pause(accessToken, deviceId = null) {
  try {
    await axios.put(
      `https://api.spotify.com/v1/me/player/pause${deviceId ? `?device_id=${deviceId}` : ''}`,
      {},
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return true;
  } catch (error) {
    console.error('Error pausing playback:', error.response?.data || error.message);
    throw error;
  }
}

export async function skipToNext(accessToken, deviceId = null) {
  try {
    await axios.post(
      `https://api.spotify.com/v1/me/player/next${deviceId ? `?device_id=${deviceId}` : ''}`,
      {},
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return true;
  } catch (error) {
    console.error('Error skipping to next:', error.response?.data || error.message);
    throw error;
  }
}

export async function skipToPrevious(accessToken, deviceId = null) {
  try {
    await axios.post(
      `https://api.spotify.com/v1/me/player/previous${deviceId ? `?device_id=${deviceId}` : ''}`,
      {},
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return true;
  } catch (error) {
    console.error('Error skipping to previous:', error.response?.data || error.message);
    throw error;
  }
}

export async function seek(accessToken, positionMs, deviceId = null) {
  try {
    await axios.put(
      `https://api.spotify.com/v1/me/player/seek`,
      null,
      {
        params: {
          position_ms: positionMs,
          ...(deviceId && { device_id: deviceId })
        },
        headers: {
          'Authorization': `Bearer ${accessToken}`,
          'Content-Type': 'application/json'
        }
      }
    );

    return true;
  } catch (error) {
    console.error('Error seeking:', error.response?.data || error.message);
    throw error;
  }
}

export async function getCurrentPlayback(accessToken) {
  try {
    const response = await axios.get(
      'https://api.spotify.com/v1/me/player',
      {
        headers: {
          'Authorization': `Bearer ${accessToken}`
        }
      }
    );

    return response.data;
  } catch (error) {
    if (error.response?.status === 204) {
      return null; // No active playback
    }
    console.error('Error getting current playback:', error.response?.data || error.message);
    throw error;
  }
}

export async function searchTracks(accessToken, query, limit = 20) {
  try {
    const response = await axios.get(
      'https://api.spotify.com/v1/search',
      {
        params: {
          q: query,
          type: 'track',
          limit
        },
        headers: {
          'Authorization': `Bearer ${accessToken}`
        }
      }
    );

    return response.data.tracks.items.map(track => ({
      uri: track.uri,
      id: track.id,
      name: track.name,
      artists: track.artists.map(a => a.name).join(', '),
      album: track.album.name,
      albumArt: track.album.images[0]?.url,
      durationMs: track.duration_ms,
      previewUrl: track.preview_url
    }));
  } catch (error) {
    console.error('Error searching tracks:', error.response?.data || error.message);
    throw error;
  }
}
